addons21/1346635841/config.md
* Color values can be either HTML Hex Codes (e.g. "#00FF00") or Color Names (e.g. "green"). There are some websites and online tools for reference, such as:
    - [Wikipedia - Web colors](https://en.wikipedia.org/wiki/Web_colors)
	- [W3Schools - Colors Tutorial](https://www.w3schools.com/colors/default.asp)


* **0101-User-Agent**: There is a default value, but it is **highly recommended** that you paste your own user agent here. You can find it by:
    - Googling "my user agent"
    - Using websites like [whatismybrowser.com](https://www.whatismybrowser.com/detect/what-is-my-user-agent) or [whatsmyua.info](https://www.whatsmyua.info/)


* **0201-Display 'Add Pronunciation' button?**: "Yes" if you want to see the "Add Pronunciation" button, "No" otherwise
* **0202-Display 'Add 1st Definition' button?**: "Yes" if you want to see the "Add 1st Definition" button, "No" otherwise
* **0203-Display 'Add All Definitions' button?**: "Yes" if you want to see the "Add All Definitions" button, "No" otherwise
* **0204-Display 'Add Translation' button?**: "Yes" if you want to see the "Add Translation" button, "No" otherwise
* **0205-Add labels to play buttons?**: "Yes" if you want the US or GB labels to appear on the respective media play buttons, "No" otherwise


* **0301-Pronunciation field number**: Field number that you want the pronunciations to be added to; typically "1" if you want to add it to the entry (word)
* **0302-Add US pronunciation?**: "Yes" if you want to add US (American) pronunciation, "No" otherwise
* **0303-Add GB pronunciation?**: "Yes" if you want to add GB (British) pronunciation, "No" otherwise
* **0304-US or GB pronunciation first?**: "GB" if you want the sequence of pronunciation files to be 1-GB 2-US, "US" if vice versa
* **0305-Keep pronunciation duplicates?**: "No" if you do not want to add the pronunciation if it already exists, "Yes" otherwise


* **0401-Not overwrite separator**: The separator with the previous contents of a field if "Overwrite 1st Definition" and/or "Overwrite All Definitions" are set to "No"; &lt;br&gt; means a line break (like pressing the Enter button on the keyboard while typing)
* **0402-Not overwrite separator color**: "Not overwrite separator" color
* **0403-Titles color**: The titles color; applies to all titles in the 1st Definition and All Definitions.
* **0404-Synonyms and antonyms separator**: The separator between the synonyms and the antonyms words
* **0405-Overwrite 1st Definition**: "Yes" if you want the definition elements to replace the previous contents of the specified fields, "No" if you want the elements to be added to them; applies to all the specified fields.
* **0406-Overwrite All Definitions**: "Yes" if you want the definitions to replace the previous contents of the specified field, "No" if you want the elements to be added to it


* **0501-1st Definition phonetic field number**: The field number to add to or replace its contents with the phonetic. Set to "0" if you want the phonetic not to appear in any field.
* **0502-1st Definition phonetic title"**: The title to add before the 1st Definition phonetic
* **0503-1st Definition phonetic color**: The 1st Definition phonetic color

* **0511-1st Definition part of speech field number**: The field number to add to or replace its contents with the part of speech. Set to "0" if you want the part of speech not to appear in any field.
* **0512-1st Definition part of speech title**: The title to add before the 1st Definition part of speech
* **0513-1st Definition part of speech color**: The 1st Definition part of speech color

* **0521-1st Definition definition field number**: The field number to add to or replace its contents with the definition. Set to "0" if you want the definition not to appear in any field.
* **0522-1st Definition definition title**: The title to add before the 1st Definition definition
* **0523-1st Definition definition color**: The 1st Definition definition color

* **0531-1st Definition example field number**: The field number to add to or replace its contents with the example. Set to "0" if you want the example not to appear in any field.
* **0532-1st Definition example title**: The title to add before the 1st Definition example
* **0533-1st Definition example color**: The 1st Definition example color

* **0541-1st Definition synonyms field number**: The field number to add to or replace its contents with the synonyms. Set to "0" if you want the synonyms not to appear in any field.
* **0542-1st Definition synonyms title**: The title to add before the 1st Definition synonyms
* **0543-1st Definition synonyms color**: The 1st Definition synonyms color. Applies to both synonyms and their separator ("Synonyms and antonyms separator").

* **0551-1st Definition antonyms field number**: The field number to add to or replace its contents with the antonyms. Set to "0" if you want the antonyms not to appear in any field.
* **0552-1st Definition antonyms title**: The title to add before the 1st Definition antonyms
* **0553-1st Definition antonyms color**: The 1st Definition antonyms color. Applies to both antonyms and their separator ("Synonyms and antonyms separator").


* **0601-All Definitions field number**: The field number to add to or replace its contents with the definitions. It overrides all the defined elements field numbers, except those who are set to "0" which means they will not appear in the definitions.

* **0611-All Definitions phonetic field number**: Phonetic will not appear if this is set to "0". Values other than "0" are overridden by "All Definitions field number".
* **0612-All Definitions phonetic title**: The title to add before the All Definitions phonetic
* **0613-All Definitions phonetic color**: All Definitions phonetic color

* **0621-All Definitions part of speech field number**: Part of speech will not appear if this is set to "0". Values other than "0" are overridden by "All Definitions field number".
* **0622-All Definitions part of speech title**: The title to add before the All Definitions part of speech
* **0623-All Definitions part of speech color**: All Definitions part of speech color

* **0631-All Definitions definition field number**: Definition will not appear if this is set to "0". Values other than "0" are overridden by "All Definitions field number".
* **0632-All Definitions definition title**: The title to add before the All Definitions definitions
* **0633-All Definitions definition color**: All Definitions definition color

* **0641-All Definitions example field number**: Example will not appear if this is set to "0". Values other than "0" are overridden by "All Definitions field number".
* **0642-All Definitions example title**: The title to add before the All Definitions example
* **0643-All Definitions example color**: All Definitions example color

* **0651-All Definitions synonyms field number**: Synonyms will not appear if this is set to "0". Values other than "0" are overridden by "All Definitions field number".
* **0652-All Definitions synonyms title**: The title to add before the All Definitions synonyms
* **0653-All Definitions synonyms color**: All Definitions synonyms color. Applies to both synonyms and their separator ("Synonyms and antonyms separator").

* **0661-All Definitions antonyms field number**: Antonyms will not appear if this is set to "0". Values other than "0" are overridden by "All Definitions field number".
* **0662-All Definitions antonyms title**: The title to add before the All Definitions antonyms
* **0663-All Definitions antonyms color**: All Definitions antonyms color. Applies to both antonyms and their separator ("Synonyms and antonyms separator").


* **0701-Translation field number**: The field number to add to or replace its contents with the translation.
* **0702-Translation target language**: Choose a target language ISO-639-1 two-letter code. You can find the code in [Google documentation](https://cloud.google.com/translate/docs/languages) or [Wikipedia](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes). Source language is always English.
* **0703-Add language name to translation title?**: Set "Yes" if you want to add the language name before the specified "Translation title". For example if "Translation target language" is set to "fr" and "Translation title" is set to " translation: " and "Add language name to translation title?" is set to "Yes", the final title shown would be "French translation: ".
* **0704-Translation title**: The title to add before the translation. It can precede by the target language name if "Add language name to translation title?" is set to "Yes". Leaving a blank value ("") will make "Add language name to translation title?" ineffective and no title will be shown.
* **0705-Translation color**: The translation color
* **0706-Add transliteration to translation?**: "Yes" if you want the transliteration (in the target language) to appear alongside the translation. Transliteration is the translation written in the Latin alphabet; for example, if the English word entry is "night", Russian translation is "ночь" and transliteration is "noch'", that helps to know how the translation is pronounced in the target language.
* **0707-Transliteration color**: The transliteration color
* **0708-Overwrite translation?**: "Yes" if you want the translation to replace the previous contents of the specified field, "No" if you want the translation to be added to it

addons21/1346635841/settings.py
# Reads the user config file (config.json), generates patterns for
# text search and defines the constants used in other modules.

from aqt import mw
import csv
import os


# Redaing the user settings in config.json
class config_values:
    def __init__(self):
        config = mw.addonManager.getConfig(__name__)
        self.mybrowser_headers = {
            'User-Agent': config['0101-User-Agent']}

        self.display_pronunciation_button = (
            config['0201-Display \'Add Pronunciation\' button?'][0].lower() == 'y')
        self.display_1st_definition_button = (
            config['0202-Display \'Add 1st Definition\' button?'][0].lower() == 'y')
        self.display_all_definitions_button = (
            config['0203-Display \'Add All Definitions\' button?'][0].lower() == 'y')
        self.display_translation_button = (
            config['0204-Display \'Add Translation\' button?'][0].lower() == 'y')
        self.add_labels_to_play_buttons = (
            config['0205-Add labels to play buttons?'][0].lower() == 'y')

        self.pronunciation_field = (
            int(config['0301-Pronunciation field number'][0].strip()) - 1)
        self.pronunciation_field_us = (
            int(config['03011-Pronunciation US field number'][0].strip()) - 1)
        self.pronunciation_field_gb = (
            int(config['03012-Pronunciation GB field number'][0].strip()) - 1)
        self.add_US_pronunciation = (
            config['0302-Add US pronunciation?'][0].lower() == 'y')
        self.add_GB_pronunciation = (
            config['0303-Add GB pronunciation?'][0].lower() == 'y')
        self.US_pronunciation_first = (
            config['0304-US or GB pronunciation first?'][0].lower() == 'u')
        self.keep_pronunciation_duplicates = (
            config['0305-Keep pronunciation duplicates?'][0].lower() == 'y')

        self.not_overwrite_separator = config['0401-Not overwrite separator']
        self.not_overwrite_separator_color = config['0402-Not overwrite separator color'].strip(
        ).lower()
        self.titles_color = config['0403-Titles color'].strip().lower()
        self.synonyms_and_antonyms_separator = config['0404-Synonyms and antonyms separator']
        self.overwrite_1st_definition = (
            config['0405-Overwrite 1st Definition?'][0].lower() == 'y')
        self.overwrite_all_definitions = (
            config['0406-Overwrite All Definitions?'][0].lower() == 'y')

        self.first_definition_phonetic_field = int(
            config['0501-1st Definition phonetic field number'].strip())-1
        self.first_definition_phonetic_title = config['0502-1st Definition phonetic title']
        self.first_definition_phonetic_color = config['0503-1st Definition phonetic color'].strip(
        ).lower()

        self.first_definitions_pos_field = int(
            config['0511-1st Definition part of speech field number'].strip())-1
        self.first_definitions_pos_title = config['0512-1st Definition part of speech title']
        self.first_definitions_pos_color = config['0513-1st Definition part of speech color'].strip(
        ).lower()

        self.first_definition_definition_field = int(
            config['0521-1st Definition definition field number'].strip())-1
        self.first_definition_definition_title = config['0522-1st Definition definition title']
        self.first_definition_definition_color = config['0523-1st Definition definition color'].strip(
        ).lower()

        self.first_definition_example_field = int(
            config['0531-1st Definition example field number'].strip())-1
        self.first_definition_example_title = config['0532-1st Definition example title']
        self.first_definition_example_color = config['0533-1st Definition example color'].strip(
        ).lower()

        self.first_definition_synonyms_field = int(
            config['0541-1st Definition synonyms field number'].strip())-1
        self.first_definition_synonyms_title = config['0542-1st Definition synonyms title']
        self.first_definition_synonyms_color = config['0543-1st Definition synonyms color'].strip(
        ).lower()

        self.first_definition_antonyms_field = int(
            config['0551-1st Definition antonyms field number'].strip())-1
        self.first_definition_antonyms_title = config['0552-1st Definition antonyms title']
        self.first_definition_antonyms_color = config['0553-1st Definition antonyms color'].strip(
        ).lower()

        self.all_definitions_field = int(
            config['0601-All Definitions field number'])-1

        self.all_definitions_phonetic_field = int(
            config['0611-All Definitions phonetic field number'].strip())-1
        self.all_definitions_phonetic_title = config['0612-All Definitions phonetic title']
        self.all_definitions_phonetic_color = config['0613-All Definitions phonetic color'].strip(
        ).lower()

        self.all_definitions_pos_field = int(
            config['0621-All Definitions part of speech field number'].strip())-1
        self.all_definitions_pos_title = config['0622-All Definitions part of speech title']
        self.all_definitions_pos_color = config['0623-All Definitions part of speech color'].strip(
        ).lower()

        self.all_definitions_definition_field = int(
            config['0631-All Definitions definition field number'].strip())-1
        self.all_definitions_definition_title = config['0632-All Definitions definition title']
        self.all_definitions_definition_color = config['0633-All Definitions definition color'].strip(
        ).lower()

        self.all_definitions_example_field = int(
            config['0641-All Definitions example field number'].strip())-1
        self.all_definitions_example_title = config['0642-All Definitions example title']
        self.all_definitions_example_color = config['0643-All Definitions example color'].strip(
        ).lower()

        self.all_definitions_synonyms_field = int(
            config['0651-All Definitions synonyms field number'].strip())-1
        self.all_definitions_synonyms_title = config['0652-All Definitions synonyms title']
        self.all_definitions_synonyms_color = config['0653-All Definitions synonyms color'].strip(
        ).lower()

        self.all_definitions_antonyms_field = int(
            config['0661-All Definitions antonyms field number'].strip())-1
        self.all_definitions_antonyms_title = config['0662-All Definitions antonyms title']
        self.all_definitions_antonyms_color = config['0663-All Definitions antonyms color'].strip(
        ).lower()

        self.translation_field = (
            int(config['0701-Translation field number'])-1)
        self.translation_target_language = config['0702-Translation target language'].strip(
        ).lower()
        self.add_language_name = (
            config['0703-Add language name to translation title?'][0].lower() == 'y')
        self.translation_title = config['0704-Translation title']
        self.translation_color = config['0705-Translation color'].strip().lower()
        self.add_transliteration = (
            config['0706-Add transliteration to translation?'][0].lower() == 'y')
        self.transliteration_color = config['0707-Transliteration color'].strip(
        ).lower()
        self.overwrite_translation = (
            config['0708-Overwrite translation?'][0].lower() == 'y')


class patterns:
    '''Generates regex or search patterns, accepting "word" as class argument.'''

    def __init__(self, word: str):
        self.word = word
        self.US_mp3_filename_pattern = f'{word}([a-zA-Z0-9._-]*)[_-]us[_-]([a-zA-Z0-9._-]*)\.mp3'
        self.GB_mp3_filename_pattern = f'{word}([a-zA-Z0-9._-]*)[_-]gb[_-]([a-zA-Z0-9._-]*)\.mp3'
        self.mp3_filename_pattern = f'{word}([a-zA-Z0-9._-]*)[_-](us|gb)[_-]([a-zA-Z0-9._-]*)\.mp3'
        self.google_pronounce_search_url = f'https://www.google.com/search?q=how+to+pronounce+{word}'
        self.google_define_search_url = f'https://www.google.com/search?q=define:{word}'


# Constant definitions
MP3_URL_BEGIN_STR = '//ssl.gstatic.com/dictionary/static/'
MP3_URL_END_STR = '.mp3'

# Loading ISO-639-1 codes / language name mapping into a python dict
_path = os.path.join(os.path.dirname(__file__), 'files/ISO-639-1_Codes.csv')
_iso_639_1_codes_file = open(_path, mode='r')
_iso_639_1_codes_dictreader = csv.DictReader(_iso_639_1_codes_file)

iso_639_1_codes_dict: dict = {}
for _row in _iso_639_1_codes_dictreader:
    iso_639_1_codes_dict[_row['ISO-639-1 Code']] = _row['Language']

_iso_639_1_codes_file.close()
